File: meowdis/src/__tests__/translator/lists.test.ts
import { describe, expect, it } from "vitest";
import { translate } from "../../translator";
import type { ResultRow } from "../../translator/types";

function expectIndexQuoted(sql: string): void {
  expect(sql).toContain('"index"');
  const stripped = sql.replaceAll('"index"', "");
  expect(stripped.includes("index")).toBe(false);
}

describe("lists translator", () => {
  it("LPUSH returns count", () => {
    const translation = translate(["LPUSH", "k", "a", "b"]);
    expect(translation.statements).toHaveLength(6);
    const results: ResultRow[][] = [[], [], [], [], [], [{ count: "2" }]];
    expect(translation.mapResult(results)).toBe(2);
  });

  it("RPUSH returns count", () => {
    const translation = translate(["RPUSH", "k", "a"]);
    expect(translation.statements).toHaveLength(5);
    const results: ResultRow[][] = [[], [], [], [], [{ count: "1" }]];
    expect(translation.mapResult(results)).toBe(1);
  });

  it("LPOP uses quoted index", () => {
    const translation = translate(["LPOP", "k"]);
    expectIndexQuoted(translation.statements[2].sql);
    const results: ResultRow[][] = [[], [], [{ value: "v" }], []];
    expect(translation.mapResult(results)).toBe("v");
  });

  it("RPOP uses quoted index", () => {
    const translation = translate(["RPOP", "k"]);
    expectIndexQuoted(translation.statements[2].sql);
    const results: ResultRow[][] = [[], [], [{ value: "v" }], []];
    expect(translation.mapResult(results)).toBe("v");
  });

  it("LRANGE handles positive and negative indices", () => {
    const translation = translate(["LRANGE", "k", "1", "2"]);
    const results: ResultRow[][] = [
      [],
      [],
      [{ value: "a" }, { value: "b" }, { value: "c" }, { value: "d" }],
    ];
    expect(translation.mapResult(results)).toEqual(["b", "c"]);

    const negative = translate(["LRANGE", "k", "-2", "-1"]);
    expect(negative.mapResult(results)).toEqual(["c", "d"]);
  });
});

File: compute-go/translator/lists_test.go
package translator

import (
	"strings"
	"testing"
)

func TestTranslateListCounts(t *testing.T) {
	lpush, err := Translate([]string{"LPUSH", "k", "a", "b"})
	if err != nil {
		t.Fatalf("Translate error: %v", err)
	}
	if len(lpush.Statements) != 6 {
		t.Fatalf("expected 6 statements, got %d", len(lpush.Statements))
	}
	results := make([][]map[string]any, 6)
	results[5] = []map[string]any{{"count": "2"}}
	value, err := lpush.MapResult(results)
	if err != nil {
		t.Fatalf("MapResult error: %v", err)
	}
	if value != int64(2) {
		t.Fatalf("expected 2, got %v", value)
	}

	rpush, err := Translate([]string{"RPUSH", "k", "a"})
	if err != nil {
		t.Fatalf("Translate error: %v", err)
	}
	if len(rpush.Statements) != 5 {
		t.Fatalf("expected 5 statements, got %d", len(rpush.Statements))
	}
	results = make([][]map[string]any, 5)
	results[4] = []map[string]any{{"count": "1"}}
	value, err = rpush.MapResult(results)
	if err != nil {
		t.Fatalf("MapResult error: %v", err)
	}
	if value != int64(1) {
		t.Fatalf("expected 1, got %v", value)
	}
}

func TestTranslatePopUsesQuotedIndex(t *testing.T) {
	lpop, err := Translate([]string{"LPOP", "k"})
	if err != nil {
		t.Fatalf("Translate error: %v", err)
	}
	if !hasQuotedIndex(lpop.Statements[2].SQL) {
		t.Fatalf("expected quoted index in LPOP SQL")
	}
	results := make([][]map[string]any, 4)
	results[2] = []map[string]any{{"value": "v"}}
	value, err := lpop.MapResult(results)
	if err != nil {
		t.Fatalf("MapResult error: %v", err)
	}
	if value != "v" {
		t.Fatalf("expected v, got %v", value)
	}

	rpop, err := Translate([]string{"RPOP", "k"})
	if err != nil {
		t.Fatalf("Translate error: %v", err)
	}
	if !hasQuotedIndex(rpop.Statements[2].SQL) {
		t.Fatalf("expected quoted index in RPOP SQL")
	}
}

func TestTranslateLRange(t *testing.T) {
	translation, err := Translate([]string{"LRANGE", "k", "1", "2"})
	if err != nil {
		t.Fatalf("Translate error: %v", err)
	}
	results := make([][]map[string]any, 3)
	results[2] = []map[string]any{{"value": "a"}, {"value": "b"}, {"value": "c"}, {"value": "d"}}
	value, err := translation.MapResult(results)
	if err != nil {
		t.Fatalf("MapResult error: %v", err)
	}
	list := value.([]any)
	if len(list) != 2 || list[0] != "b" || list[1] != "c" {
		t.Fatalf("unexpected LRANGE result: %v", list)
	}

	negative, err := Translate([]string{"LRANGE", "k", "-2", "-1"})
	if err != nil {
		t.Fatalf("Translate error: %v", err)
	}
	value, err = negative.MapResult(results)
	if err != nil {
		t.Fatalf("MapResult error: %v", err)
	}
	list = value.([]any)
	if len(list) != 2 || list[0] != "c" || list[1] != "d" {
		t.Fatalf("unexpected LRANGE result: %v", list)
	}
}

func hasQuotedIndex(sql string) bool {
	if sql == "" {
		return false
	}
	stripped := replaceAll(sql, "\"index\"", "")
	return !strings.Contains(stripped, "index")
}

func replaceAll(value, old, new string) string {
	for {
		idx := strings.Index(value, old)
		if idx < 0 {
			return value
		}
		value = value[:idx] + new + value[idx+len(old):]
	}
}
